working-with-collections/main.js

```diff
@@ -1,6 +1,105 @@
+/* eslint-disable no-prototype-builtins */
 console.log('Lodash is loaded:', typeof _ !== 'undefined');
 
-// var player1 = {
-//   name: 'name1',
-//   hand: []
-// };
+function createCards() {
+  var cards = [];
+  var ranks = ['Ace', 2, 3, 4, 5, 6, 7, 8, 9, 10, 'Jack', 'Queen', 'King'];
+  var suits = ['diamonds', 'clubs', 'hearts', 'spades'];
+  for (var i = 0; i < 52; i++) {
+    var card = {};
+    card.rank = ranks[i % 13];
+    card.suit = suits[Math.floor(i / 13)];
+    cards.push(card);
+  }
+  return cards;
+}
+
+function createPlayers(numPlayers) {
+  var players = [];
+  for (var i = 0; i < numPlayers; i++) {
+    var player = {};
+    player.name = 'Player ' + (i + 1);
+    player.hand = [];
+    players.push(player);
+  }
+  return players;
+}
+
+function shuffle(totalCards) {
+  while (cardIndexes.length <= totalCards) {
+    var cardIndex = Math.floor(Math.random() * 52);
+    if (!cardIndexes.includes(cardIndex)) {
+      cardIndexes.push(cardIndex);
+    }
+  }
+}
+
+function deal(totalCards, startIndex) {
+  for (var i = startIndex; i < totalCards; i++) {
+    var card = cards[cardIndexes[i]];
+    players[i % players.length].hand.push(card);
+  }
+}
+
+function score() {
+  for (var i = 0; i < players.length; i++) {
+    var score = 0;
+    var name = players[i].name;
+    var hand = players[i].hand;
+    for (var j = 0; j < hand.length; j++) {
+      var card = hand[j];
+      if (card.rank === 'Jack' || card.rank === 'Queen' || card.rank === 'King') {
+        score += 10;
+      } else if (card.rank === 'Ace') {
+        score += 11;
+      } else {
+        score += card.rank;
+      }
+    }
+    if (score > highScore) {
+      winners = [name];
+      highScore = score;
+    } else if (score === highScore) {
+      winners.push(name);
+    }
+  }
+}
+
+function playGame(players, numCardsPerHand) {
+  var totalCards = players.length * numCardsPerHand;
+  shuffle(totalCards);
+  deal(totalCards, 0);
+  score();
+  if (winners.length > 1) {
+    handleTie(totalCards);
+  }
+  var winnerName = winners[0];
+  console.log('The winner is', winnerName + '!');
+  console.log('Players:', players);
+}
+
+function handleTie(totalCards) {
+  while (winners.length > 1) {
+    console.log('Tie - deal another card as tie-breaker');
+    var tiePlayers = [];
+    for (var i = 0; i < players.length; i++) {
+      if (winners.includes(players[i].name)) {
+        tiePlayers.push(players[i]);
+      }
+    }
+    players = tiePlayers;
+    var dealStartIndex = totalCards;
+    totalCards += winners.length;
+    shuffle(totalCards);
+    deal(totalCards, dealStartIndex);
+    winners = [];
+    score();
+  }
+}
```

Comment:
Try adjusting your algorithm so that most of your functions `return` a value instead of modifying existing global values. You should be able to only have 3 global variables:
- create the players
- create the card deck
- set the number of cards per hand

```js
function playGame(players, cardDeck, cardsPerHand) {
  // create a shuffled deck
  // deal the cards to the players
  // get an array of winners
  // if tie:
  //   empty hand of each winner
  //   play game again with the winners this time <- recursion
  // else:
  //   declare winner
}
```

A recursive `playGame` function would be very elegant.

https://en.wikipedia.org/wiki/Recursion (see In Computer Science).

Reply:
Ahh okay, I see! I decided to create the global variables only because I was thinking that we would not want to reshuffle the cards in the case of a tie, so I had to save the cards that were already dealt. I'll see if I can figure out how to rewrite this with recursion! 😄